### ISSUE TITLE
Counter bug in w2v forced alignment tutorial

### ISSUE THREAD

**Issue by rbracco** [pytorch/audio#3439]
### 🐛 Describe the bug

Hey @mthrok, first off thanks for the [Forced Alignment Tutorial](https://pytorch.org/audio/0.11.0/tutorials/forced_alignment_tutorial.html), it was really fun to play around with. I believe I found that the `backtrack` step has a subtle bug that can lead to `Point` objects being assigned to the wrong `token_index`. E.G. if the label we're aligning is `me` and the first timestep that is aligned to `e` has a larger score than any of the timesteps assigned to `m`, then that timestep will be added to the path, but with token index 0 (m) instead of 1 (e). This primarily causes problems at word boundaries because the probability scores for the space-separator token '|' are often near 1. 

### Current code
```python
# Main loop of backtrack step
for t in range(t_start, 0, -1):
        # 1. Figure out if the current position was stay or change
        # Note (again):
        # `emission[J-1]` is the emission at time frame `J` of trellis dimension.
        # Score for token staying the same from time frame J-1 to T.
       stayed = trellis[t - 1, j] + probs[t - 1, blank_idx]
        # Score for token changing from C-1 at T-1 to J at T.
        changed = trellis[t - 1, j - 1] + probs[t - 1, tokens[j - 1]]

        # 2. Store the path with frame-wise probability.
         prob = probs[t - 1, tokens[j - 1] if changed > stayed else 0].exp().item()

        # Return token index and time index in non-trellis coordinate.
        # BUG HERE: path is appended with token_index `j-1` but it should be done 
        # after updating the token in step 3. Fixed code below
        path.append(Point(j - 1, t - 1, prob))

        # 3. Update the token
        if changed > stayed:
            j -= 1
            if j == 0:
                break
    else:
        raise ValueError("Failed to align")
    return path[::-1]
```

### Fixed code
```python
# Main loop of backtrack step
for t in range(t_start, 0, -1):
        # 1. Figure out if the current position was stay or change
        # Note (again):
        # `emission[J-1]` is the emission at time frame `J` of trellis dimension.
        # Score for token staying the same from time frame J-1 to T.
       stayed = trellis[t - 1, j] + probs[t - 1, blank_idx]
        # Score for token changing from C-1 at T-1 to J at T.
        changed = trellis[t - 1, j - 1] + probs[t - 1, tokens[j - 1]]

        # 2. Store the path with frame-wise probability.
         prob = probs[t - 1, tokens[j - 1] if changed > stayed else 0].exp().item()

        # 3. Update the token
        if changed > stayed:
            j -= 1

        # Fix: Only append to path after token_index j has been updated. Move break to bottom of loop.
        path.append(Point(j - 1, t - 1, prob))

         if j == 0:
             break
    else:
        raise ValueError("Failed to align")
    return path[::-1]
```

### Versions

torchaudio 0.13.1
torch 1.13.1

**Comment by mthrok** [comment]
Hi @rbracco 

I think you are right. I also think `path` array should be initialized with one element for the final position, before this loop starts. Also the visualization right after this seems to be off-by-one for both axes as it uses emission coordinate.

BTW, this week, all sudden I am getting many inquiries about FA tutorial. Do you know why? Was it featured somewhere?

**Comment by rbracco** [comment]
Yeah sorry, I should've waited to open an issue, I thought I'd pinned it down but now I'm having trouble wrapping my head around it as well. In my code I made some changes to the trellis generation as well so I'm worried I might have broken something. I'm stepping through the colab now trying to check. I'll report back if I find anything

In my case a friend linked me in response to a question I'd asked about force alignment.

**Comment by mthrok** [comment]
> Yeah sorry, I should've waited to open an issue, I thought I'd pinned it down but now I'm having trouble wrapping my head around it as well. In my code I made some changes to the trellis generation as well so I'm worried I might have broken something. I'm stepping through the colab now trying to check. I'll report back if I find anything

I tried your suggestion in #3440
https://docs-preview.pytorch.org/pytorch/audio/3440/index.html (will be available once build_docs CI job is done.)

<img width="731" alt="Screenshot 2023-06-13 at 3 11 26 PM" src="https://github.com/pytorch/audio/assets/855818/1104b983-3bf7-45fc-97cb-154472dee20a">

<img width="819" alt="Screenshot 2023-06-13 at 3 14 18 PM" src="https://github.com/pytorch/audio/assets/855818/dfe1bbe5-2ec8-4266-8004-bbddc1f486bc">

The average word-wise probability is lower. Let me think again.

> 
> In my case a friend linked me in response to a question I'd asked about force alignment.
> 
> Also did the trellis code change? This section of the get_trellis code in colab is different than what I remember
> 
> ```
> # Trellis has extra diemsions for both time axis and tokens.
>     # The extra dim for tokens represents <SoS> (start-of-sentence)
>     # The extra dim for time axis is for simplification of the code.
>     trellis = torch.empty((num_frame + 1, num_tokens + 1))
>     trellis[0, 0] = 0
>     trellis[1:, 0] = torch.cumsum(emission[:, blank_id], 0)
>     trellis[0, -num_tokens:] = -float("inf")
>     trellis[-num_tokens:, 0] = float("inf")
> ```

Few changes were made such as #3172 and  #2544.

**Comment by rbracco** [comment]
Thanks, I'll take a look again tomorrow and let you know if I make any progress.

**Comment by rbracco** [comment]
Sorry for the premature issue, I was able to resolve it. 

This was the issue:
`prob = probs[t - 1, tokens[j - 1] if changed > stayed else 0].exp().item()`, should be `prob = probs[t - 1, tokens[j - 1] if changed > stayed else blank_id].exp().item()`

This only affects implementations where `blank_id` is not 0 so code in the tutorial is not affected. Thanks for helping me to debug.

**Comment by mthrok** [comment]
@rbracco Thanks for the follow-up. I remembered that the code in tutorial is somewhat simplified version for educational purpose, and it is known that the treatment of blank tokens is slightly different than what it should be. 

You can look at more accurate version https://github.com/pytorch/audio/pull/3067/files#diff-de50e62569b0186b8b6cd648b3be81585132cbea26be2b61880876f2194a0e2e, which is the basis of the new C++ implementation https://pytorch.org/audio/main/generated/torchaudio.functional.forced_align.html#torchaudio.functional.forced_align, which works on CPU/GPU and can do a bit more elaborated things
https://pytorch.org/audio/main/tutorials/ctc_forced_alignment_api_tutorial.html#dealing-with-missing-transcripts-using-the-star-token

**Comment by rbracco** [comment]
Thank you @mthrok that clarifies things a lot. I had noticed that the final token could only ever have a duration of 1 timestep in the best path which was suboptimal, but adding a space token to act as <SOS> and <EOS> fixes that. 

Please note in the linked PR the added '|' is mistakenly referred to as a blank instead of a space. Also I think inserting a blank as SOS/EOS won't work, but adding a space does.

**Comment by mthrok** [comment]
> Please note in the linked PR the added '|' is mistakenly referred to as a blank instead of a space. Also I think inserting a blank as SOS/EOS won't work, but adding a space does.

Thanks for pointing out. Fixed.
